### PR TITLE
Guarin lig 599 add dinoexamples

### PR DESCRIPTION
**Pull request by guarin**
Adds examples for dino. Fixes some small issues with other pytorch lightning models.

### CODE REVIEW

File: examples/pytorch/dino.py [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+        y = self.teacher_backbone(x).flatten(start_dim=1)
+        z = self.teacher_head(y)
+        return z
+
+resnet = torchvision.models.resnet18()
+backbone = nn.Sequential(*list(resnet.children())[:-1])
+input_dim = 512
+# instead of a resnet you can also use a vision transformer backbone as in the
+# original paper (you might have to reduce the batch size in this case):
+# backbone = torch.hub.load('facebookresearch/dino:main', 'dino_vits16', pretrained=False)
```
Comment (IgorSusmelj) [inline examples/pytorch/dino.py:38]: Would add an example using timm models (see https://github.com/rwightman/pytorch-image-models) here in the comments. Either additionally or instead of the facebook research one. Just simply because the repo has tons of more models and is well maintained :)

Comment (guarin) [reply examples/pytorch/dino.py:38]: I just tested it and the vision transformers in timm expect all input images to have a fixed size. But DINO makes global and local crops with various sizes :(

File: examples/pytorch/dino.py [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+        views = [view.to(device) for view in views]
+        global_views = views[:2]
+        teacher_out = [model.forward_teacher(view) for view in global_views]
+        student_out = [model.forward(view) for view in views]
+        loss = criterion(teacher_out, student_out, epoch=epoch)
+        total_loss += loss.detach()
+        loss.backward()
+        optimizer.step()
+        optimizer.zero_grad()
+        update_momentum(model.student_backbone, model.teacher_backbone, m=0.99)
```
Comment (philippmwirth) [inline examples/pytorch/dino.py:87]: here you are doing the momentum update at the end of the step and in the lightning example you do it in the beginning. can we be consistent or is there a specific reason for this?

**Review by philippmwirth — CHANGES_REQUESTED** [review]
I think you forgot to add the model to the index file as I can only access it by the direct link in the docs.

Did you try it with a vision transformer? If yes does it work? Do you have any numbers as to how good the model is if you run the example?

**Comment by guarin** [comment]
> I think you forgot to add the model to the index file as I can only access it by the direct link in the docs.
Added it :)

> Did you try it with a vision transformer?
Yes

> If yes does it work? 
The loss decreases over many epochs so it looks ok.

> Do you have any numbers as to how good the model is if you run the example?
There is no validation code in the examples. I'll update and run the benchmarks to see how well it works.

File: examples/pytorch/dino.py (hunk 2 above)
Comment (guarin) [reply examples/pytorch/dino.py]: I moved it now to the top of the training loop. In general I would prefer it to be at the bottom of the loop but this does not work for pytorch lightning. This is because the optimizer step is executed after the forward step and thus it would no make sense to update the momentum at the end of the forward step.

**Review by philippmwirth — APPROVED** [review]
lgtm 🥳
